### PR TITLE
[llvm][MC][ARM][Assembly] Emit relocations for ADRs and big-endian targets

### PR DESCRIPTION
**Pull request by eleanor-arm**
Follow-up on https://github.com/llvm/llvm-project/pull/72873/
    
    When ADR/LDR instructions reference a label in a different section, the
    offset is not known until link time, however, the assembler assumes it
    can resolve them in some cases.
    
    The previous patch addressed the issue for most LDR instructions,
    focusing on little-endian targets.
    
    This patch addresses the remaining work for ADRs and big-endian targets.

### CODE REVIEW

File: llvm/test/MC/ARM/pcrel-adr16-relocs.s [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    bx lr
+
+// Checking the encoding only, as the disassembly is not quite correct here.
+
+// Thumb16 encoding supports only adding of the encoded immediate (not
+// subtracting, see [Arm ARM]), therefore sign change is required if the pcrel
+// offset is negative. This makes the calculation of the addend for
+// R_ARM_THM_PC8 more complex, for details see [ELF for the Arm 32-bit
+// architecture].
+@ ADDEND: a0ff
```
Comment (DavidSpickett) [inline llvm/test/MC/ARM/pcrel-adr16-relocs.s:22]: How exactly is the disassembly not correct?

This check could be fooled by the number a0ff appearing anywhere in the output, so we at least need to check for something else on or immediately before that line.

Comment (eleanor-arm) [reply llvm/test/MC/ARM/pcrel-adr16-relocs.s]: I've updated it in the next version. Hopefully that works.

File: llvm/test/MC/ARM/pcrel-adr16-relocs.s [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+    .balign 4
+    .global _func1
+    .type _func1, %function
+_func1:
+    adr r0, _func2
+@ RELOC: R_ARM_THM_PC8
+    bx lr
+
+// Checking the encoding only, as the disassembly is not quite correct here.
```
Comment (DavidSpickett) [inline llvm/test/MC/ARM/pcrel-adr16-relocs.s:15]: "here, it should be..."

**Review by DavidSpickett — APPROVED** [review]
LGTM

**Comment by eleanor-arm** [comment]
Thanks for the review!

**Comment by bd1976bris** [comment]
@eleanor-arm / @smithp35 / @DavidSpickett - I don't think this patch or the proceeding one is valid. 

I'm motivated by the code added here: https://reviews.freebsd.org/D2035 to FreeBSD where `cachebailout` is referenced with `adr r3, _C_LABEL(cachebailout)` and both the reference and the symbol are in the same section (`.text`). Previously, this was codegen'd either with an `add` (if the target was at a higher address) or a `sub` (lower). 

With these patches the compiler always codegen's as if the reference was in another section to the target, even when they are in the same section. That means that the compiler can't use the information as to whether the target is at a higher or lower address to the reference, and it seems that the compiler always codegen's `adr r3, _C_LABEL(cachebailout)` as a `sub` in that case. This means that the target may is out of range (it is in the FreeBSD example) and the linker will error trying to patch the relocation for the reference. 

I think that these patches need reworking so that either relocations are not used in the case where the reference and target are in the same section, or a more complicated relocation/relaxation scheme is used to allow the linker to adjust the instruction used. Alternatively, we could replace these patches with something simpler, for example the GNU assembler seems to simply error if the target for an `adr`/`ldr` is in a different section - perhaps that would be more suitable?

**Comment by smithp35** [comment]
Apologies I'm going to be away from a work computer for the next couple of weeks due to holiday, can only leave a few comments for now. In theory the assembler fixup and the linker relocation are semantically identical, the linker is expected to convert the add into a sub and vice-versa regardless of what the object file uses (https://github.com/llvm/llvm-project/blob/main/lld/ELF/Arch/ARM.cpp#L467C25). If there is an out of range error then something else may have gone wrong. Either the assembler is not writing out the object file correctly, or the linker is not resolving the relocation correctly. Are you able to help with a reproducer either for the assembler or the linker? I think Free BSD uses lld as the linker?

I agree that it would be good to have the relocation resolved at assembly time if in the same section. From memory this wasn't easy to do for just these relocations in https://github.com/llvm/llvm-project/blob/main/llvm/lib/MC/MCAssembler.cpp#L248.  Looking at the history of this problem, there is an epic thread in https://reviews.llvm.org/D72892 which looks like it involved these same BSD source files.

While I think doing the extra work to either emit an error message or do the fixup (for just these relocations when the fixup is in the same section) is possible. Given that the assembler/linker combination should work, I'd prefer to get to the bottom of why this case is failing first.

**Comment by bd1976bris** [comment]
> Apologies I'm going to be away from a work computer for the next couple of weeks due to holiday, can only leave a few comments for now. 

Thanks for the quick reply. I am also away until the new year now so progress on this will have to wait until then.

>In theory the assembler fixup and the linker relocation are semantically identical, the linker is expected to convert the add into a sub and vice-versa regardless of what the object file uses (https://github.com/llvm/llvm-project/blob/main/lld/ELF/Arch/ARM.cpp#L467C25). If there is an out of range error then something else may have gone wrong. Either the assembler is not writing out the object file correctly, or the linker is not resolving the relocation correctly. Are you able to help with a reproducer either for the assembler or the linker? I think Free BSD uses lld as the linker?

Yes, I believe that FreeBSD supports linking with LLD; I'm unsure as to whether they also support the GNU linkers? @emaste might know more? 

Thanks for the clarification on the relocation behaviour. I'm afraid that my arm/thumb knowledge is very rusty indeed and I didn't spot this in LLD ( was also in a bit of a rush before the holidays) - apologies.

> I agree that it would be good to have the relocation resolved at assembly time if in the same section. From memory this wasn't easy to do for just these relocations in https://github.com/llvm/llvm-project/blob/main/llvm/lib/MC/MCAssembler.cpp#L248. Looking at the history of this problem, there is an epic thread in https://reviews.llvm.org/D72892 which looks like it involved these same BSD source files.

Thanks for the thread link. I'll read it.
 
> While I think doing the extra work to either emit an error message or do the fixup (for just these relocations when the fixup is in the same section) is possible. Given that the assembler/linker combination should work, I'd prefer to get to the bottom of why this case is failing first.

I should be able to come up with a reproducer based on that FreeBSD code. I'll work on supplying that when I get back in the new year.

**Comment by bd1976bris** [comment]
> In theory the assembler fixup and the linker relocation are semantically identical, the linker is expected to convert the add into a sub and vice-versa regardless of what the object file uses (https://github.com/llvm/llvm-project/blob/main/lld/ELF/Arch/ARM.cpp#L467C25). If there is an out of range error then something else may have gone wrong. 

Actually, I want to clarify that the error I saw reported was:

```
error: relocation R_ARM_ALU_PC_G0 cannot be used against symbol 'cachebailout'; recompile with -fPIC
>>> defined in cpu_asm-v6.S.obj
>>> referenced by cpu_asm-v6.S:(cpu_asm-v6.S)
>>>               cpu_asm-v6.S.obj:(dcache_wb_pou_checked)
```

Given that, as you pointed out, the relocation patching in LLD should be equivalent to the assembler resolving the reference this could be caused by user error on my part.

**Comment by smithp35** [comment]
Thanks for the update, I think this needs an update in LLD. Essentially letting it know that R_ARM_ALU_PC_G0 is position independent so it can be used in PIC/PIE. Should be a quick fix.

**Comment by bd1976bris** [comment]
> Thanks for the update, I think this needs an update in LLD. Essentially letting it know that R_ARM_ALU_PC_G0 is position independent so it can be used in PIC/PIE. Should be a quick fix.

Great. Reproducer is here: https://godbolt.org/z/85KzbYcsn.

**Comment by eleanor-arm** [comment]
The fix is here -  https://github.com/llvm/llvm-project/pull/77304

**Comment by bd1976bris** [comment]
@eleanor-arm, @smithp35 thanks for the fix :)

Unfortunately, our CI shows failures with other relocation types. I'm sorry for not reporting this previously. For example, in [locore-v6.S](https://svnweb.freebsd.org/base/stable/12/sys/arm/arm/locore-v6.S?view=markup) `cpu_halt` references `cpu_reset_address` defined in [machdep.c](https://svnweb.freebsd.org/base/stable/12/sys/arm/arm/machdep.c?view=markup). I am seeing a `recompile with -fPIC` error for an `R_ARM_ABS32` for that reference with your change. I think that this is a reproducer: 

```
        .text
        .arm
_start: .word dest
        .section .text.1, "ax", %progbits
.globl dest
dest: .word 10
```

see: https://godbolt.org/z/1reqb1xE6

I am also seeing similar errors for R_ARM_THM_MOVW_ABS_NC and R_ARM_THM_MOVT_ABS relocations. I'll work on getting a full set of errors out of our CI and reported to you. Although, it might be better if we could somehow identify the possible set of affected relocation types without experimentation?

**Comment by smithp35** [comment]
Looking at the compiler explorer clang for this target is passing the linker -pie by default. I think this may be the root cause of the CI failures as it looks like some position dependent code is now being linked into a position independent executable. 

The recent changes to the assembler don't affect any of these relocations I can reproduce this with clang-10 and lld-10 from ubuntu 20.04.

If your CI code is kernel code that isn't position independent then you'll need to pass -no-pie to the compiler.

From compiler explorer adding -v to the command line (see -pie)
```
/opt/compiler-explorer/clang-trunk/bin/ld.lld" --sysroot=/opt/compiler-explorer/arm/gcc-12.2.0/arm-unknown-linux-gnueabi/arm-unknown-linux-gnueabi/sysroot -EL -X --hash-style=gnu --eh-frame-hdr -m armelf_linux_eabi -pie -dynamic-linker /lib/ld-linux.so.3 -o /app/output.s -L./lib /tmp/example-9aab2b.o -rpath ./lib -rpath /opt/compiler-explorer/arm/gcc-12.2.0/arm-unknown-linux-gnueabi/lib64 -rpath /opt/compiler-explorer/arm/gcc-12.2.0/arm-unknown-linux-gnueabi/lib32
```

The example you have there is not PIE. There is a non-relative relocation in the .text segment (R_ARM_ABS32 is put the absolute address of the symbol into the .word) in the example. By default LLD will accept R_ARM_ABS32 in .data and .data.rel.ro but not .text (it will with -z notext)

R_ARM_THM_MOVW_ABS_NC and R_ARM_THM_MOVT_ABS are also non-relative relocations that can't be used in -pie or -pic. There are PC-relative equivalents R_ARM_MOVW_PREL_NC and R_ARM_THM_MOVT_PREL that can be used instead.

**Comment by bd1976bris** [comment]
> Looking at the compiler explorer clang for this target is passing the linker -pie by default. I think this may be the root cause of the CI failures as it looks like some position dependent code is now being linked into a position independent executable.

Thanks Peter. Very much appreciated. This needs further investigation from our side.

**Comment by bd1976bris** [comment]
> The example you have there is not PIE. There is a non-relative relocation in the .text segment (R_ARM_ABS32 is put the absolute address of the symbol into the .word) in the example. By default LLD will accept R_ARM_ABS32 in .data and .data.rel.ro but not .text (it will with -z notext)

@smithp35 thanks for the information on this. To update you, it seems that the problems with the ABS relocations began with https://github.sie.sony.com/SIE-Private/cpu-toolchain-ppr/commit/42e4967140e345923a43f809ba69be57200f46ae. Unfortunately, that was close enough to this change to be tested together on CI.

**Pull request merged**